ksqldb-engine/src/main/java/io/confluent/ksql/internal/StorageUtilizationMetrics.java

```diff
@@ -0,0 +1,280 @@
+      );
+      metricsSeen.put(queryId, new HashMap<>());
+    }
+    final TaskStorageMetric newMetric;
+    // We haven't seen a metric for this query's task before
+    if (!metricsSeen.get(queryId).containsKey(taskId)) {
+      // create a new task level metric to track state store storage usage
+      newMetric = new TaskStorageMetric(
+        metricRegistry.metricName(
+          "task-storage-usage",
+          METRIC_GROUP,
+          ImmutableMap.of("task-id", taskId, "query-id", queryId)
+        ));
+      // add to list of seen task metrics for this query
+      metricsSeen.get(queryId).put(taskId, newMetric);
+      // create gauge for task level storage usage
+      metricRegistry.addMetric(
+          newMetric.metricName,
+          (Gauge<BigInteger>) (config, now) -> newMetric.getValue()
+      );
+    } else {
+      // We have this metric already
+      newMetric = metricsSeen.get(queryId).get(taskId);
+    }
+    newMetric.add(metric);
+  }
+
+  private synchronized void handleRemovedSstFileSizeMetric(
+      final TaskStorageMetric taskMetric,
+      final KafkaMetric metric,
+      final String queryId,
+      final String taskId
+  ) {
+    // remove storage metric for this task
+    taskMetric.remove(metric);
+    if (taskMetric.metrics.size() == 0) {
+      // no more storage metrics for this task, can remove task gauge
+      metricRegistry.removeMetric(taskMetric.metricName);
+      metricsSeen.get(queryId).remove(taskId);
+      if (metricsSeen.get(queryId).size() == 0) {
+        // we've removed all the task metrics for this query, don't need the query metrics anymore
+        metricRegistry.removeMetric(metricRegistry.metricName(
+            "query-storage-usage",
+            METRIC_GROUP,
+            ImmutableMap.of("query-id", queryId))
+        );
+      }
+    }
+  }
+
+  private BigInteger computeQueryMetric(final String queryId) {
+    BigInteger queryMetricSum = BigInteger.ZERO;
+    for (final Supplier<BigInteger> gauge : getGaugesForQuery(queryId)) {
+      queryMetricSum = queryMetricSum.add(gauge.get());
+    }
+    return queryMetricSum;
+  }
+
+  private synchronized Collection<Supplier<BigInteger>> getGaugesForQuery(final String queryId) {
+    return metricsSeen.get(queryId).values().stream()
+      .map(v -> (Supplier<BigInteger>) v::getValue)
+      .collect(Collectors.toList());
+  }
+
+  private String getQueryId(final KafkaMetric metric) {
+    final String queryIdTag = metric.metricName().tags().getOrDefault("thread-id", "");
+    final Pattern pattern = Pattern.compile("(?<=query_|transient_)(.*?)(?=-)");
+    final Matcher matcher = pattern.matcher(queryIdTag);
+    return matcher.find() ? matcher.group(1) : "";
+  }
+
+  static void reset() {
+    registeredNodeMetrics.getAndSet(false);
+  }
+
+  private static class TaskStorageMetric {
+    final MetricName metricName;
+    private final Map<MetricName, KafkaMetric> metrics = new ConcurrentHashMap<>();
```

Comment:
Hmm, are we having more than one metric here? Since we are filtering on `metric.metricName().name().equals("total-sst-files-size")` at the first place, we would end up with only that metric name right?

Reply:
From what I understand, `total-sst-files-size` is broken down by store, so we may have multiple stores per task. The `storeName` is included in the tag map so the metric names would be different if you had multiple stores under one task

Reply:
Ah yes, you're right :)